File: src/common/ty/ty_ancestors.ml
(**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

open Ty_symbol

class ['self] iter_ty_base =
  object (_ : 'self)
    method private on_string : 'env. 'env -> string -> unit = (fun _env _x -> ())

    method private on_bool : 'env. 'env -> bool -> unit = (fun _env _x -> ())

    method private on_int : 'env. 'env -> int -> unit = (fun _env _x -> ())

    method private on_symbol : 'env. 'env -> symbol -> unit = (fun _env _x -> ())

    method private on_aloc : 'env. 'env -> ALoc.t -> unit = (fun _env _x -> ())

    method private on_option : 'env 'a 'b. ('env -> 'a -> 'b) -> 'env -> 'a option -> 'b option =
      (fun f env -> Option.map ~f:(f env))

    method private on_list : 'env 'a. ('env -> 'a -> unit) -> 'env -> 'a list -> unit =
      (fun f env -> List.iter (f env))
  end

class ['self] iter2_ty_base =
  object (self : 'self)
    method private on_string : 'env. 'env -> string -> string -> unit = (fun _env _x _y -> ())

    method private on_bool : 'env. 'env -> bool -> bool -> unit = (fun _env _x _y -> ())

    method private on_int : 'env. 'env -> int -> int -> unit = (fun _env _x _y -> ())

    method private on_symbol : 'env. 'env -> symbol -> symbol -> unit = (fun _env _x _y -> ())

    method private on_aloc : 'env. 'env -> ALoc.t -> ALoc.t -> unit = (fun _env _x _y -> ())

    method private fail_option : 'env 'a. 'env -> 'a option -> 'a option -> unit =
      (fun _ _ _ -> raise VisitorsRuntime.StructuralMismatch)

    method private on_option
        : 'env 'a. ('env -> 'a -> 'a -> unit) -> 'env -> 'a option -> 'a option -> unit =
      fun f env x y ->
        match (x, y) with
        | (Some x, Some y) -> f env x y
        | (Some _, None)
        | (None, Some _) ->
          self#fail_option env x y
        | (None, None) -> ()

    method private fail_list : 'env 'a. 'env -> 'a list -> 'a list -> unit =
      (fun _ _ _ -> raise VisitorsRuntime.StructuralMismatch)

    method private on_list
        : 'env 'a. ('env -> 'a -> 'a -> unit) -> 'env -> 'a list -> 'a list -> unit =
      fun f env l1 l2 ->
        match (l1, l2) with
        | ([], []) -> ()
        | (a1 :: l1, a2 :: l2) ->
          f env a1 a2;
          self#on_list f env l1 l2
        | (l1, l2) -> self#fail_list env l1 l2
  end

class ['self] map_ty_base =
  object (_ : 'self)
    method private on_string : 'env -> string -> string = (fun _ x -> x)

    method private on_bool : 'env -> bool -> bool = (fun _ x -> x)

    method private on_int : 'env -> int -> int = (fun _ x -> x)

    method private on_symbol : 'env -> symbol -> symbol = (fun _ x -> x)

    method private on_aloc : 'env -> ALoc.t -> ALoc.t = (fun _ x -> x)

    method private on_list : 'env 'a 'b. ('env -> 'a -> 'b) -> 'env -> 'a list -> 'b list =
      (fun f env -> Core_list.map ~f:(f env))

    method private on_option : 'env 'a 'b. ('env -> 'a -> 'b) -> 'env -> 'a option -> 'b option =
      (fun f env -> Option.map ~f:(f env))
  end

class ['self] endo_ty_base =
  object (_self : 'self)
    method private on_string : 'env -> string -> string = (fun _ x -> x)

    method private on_bool : 'env -> bool -> bool = (fun _ x -> x)

    method private on_int : 'env -> int -> int = (fun _ x -> x)

    method private on_symbol : 'env -> symbol -> symbol = (fun _ x -> x)

    method private on_aloc : 'env -> ALoc.t -> ALoc.t = (fun _ x -> x)

    (* Copied from
     * https://github.com/facebook/hhvm/blob/master/hphp/hack/src/ast/ast_defs_visitors_ancestors.ml
     *)
    method private on_list : 'env 'a. ('env -> 'a -> 'a) -> 'env -> 'a list -> 'a list =
      fun f env xs ->
        let rec aux env xs counter =
          match xs with
          | [] -> xs
          | [y1] ->
            let z1 = f env y1 in
            if y1 == z1 then
              xs
            else
              [z1]
          | [y1; y2] ->
            let z1 = f env y1 in
            let z2 = f env y2 in
            if y1 == z1 && y2 == z2 then
              xs
            else
              [z1; z2]
          | [y1; y2; y3] ->
            let z1 = f env y1 in
            let z2 = f env y2 in
            let z3 = f env y3 in
            if y1 == z1 && y2 == z2 && y3 == z3 then
              xs
            else
              [z1; z2; z3]
          | [y1; y2; y3; y4] ->
            let z1 = f env y1 in
            let z2 = f env y2 in
            let z3 = f env y3 in
            let z4 = f env y4 in
            if y1 == z1 && y2 == z2 && y3 == z3 && y4 == z4 then
              xs
            else
              [z1; z2; z3; z4]
          | [y1; y2; y3; y4; y5] ->
            let z1 = f env y1 in
            let z2 = f env y2 in
            let z3 = f env y3 in
            let z4 = f env y4 in
            let z5 = f env y5 in
            if y1 == z1 && y2 == z2 && y3 == z3 && y4 == z4 && y5 == z5 then
              xs
            else
              [z1; z2; z3; z4; z5]
          | y1 :: y2 :: y3 :: y4 :: y5 :: ys ->
            let z1 = f env y1 in
            let z2 = f env y2 in
            let z3 = f env y3 in
            let z4 = f env y4 in
            let z5 = f env y5 in
            let zs =
              if counter > 1000 then
                aux_slow env ys [] ys false
              else
                aux env ys (counter + 1)
            in
            if y1 == z1 && y2 == z2 && y3 == z3 && y4 == z4 && y5 == z5 && ys == zs then
              xs
            else
              z1 :: z2 :: z3 :: z4 :: z5 :: zs
        and aux_slow env xs acc original_list has_new_elements =
          match xs with
          | [] ->
            if has_new_elements then
              List.rev acc
            else
              original_list
          | y1 :: ys ->
            let z1 = f env y1 in
            aux_slow env ys (z1 :: acc) original_list (has_new_elements || y1 != z1)
        in
        aux env xs 0

    method private on_option : 'env 'a. ('env -> 'a -> 'a) -> 'env -> 'a option -> 'a option =
      fun f env x ->
        match x with
        | None -> x
        | Some y ->
          let z = f env y in
          if y == z then
            x
          else
            Some z
  end

class virtual ['e] monoid =
  object
    method virtual private zero : 'e

    method virtual private plus : 'e -> 'e -> 'e
  end

class virtual ['self] reduce_ty_base =
  object (self : 'self)
    inherit ['acc] monoid

    method private on_string : 'env. 'env -> string -> 'acc = (fun _ _ -> self#zero)

    method private on_int : 'env. 'env -> int -> 'acc = (fun _ _ -> self#zero)

    method private on_bool : 'env. 'env -> bool -> 'acc = (fun _ _ -> self#zero)

    method private on_symbol : 'env. 'env -> symbol -> 'acc = (fun _ _ -> self#zero)

    method private on_aloc : 'env. 'env -> ALoc.t -> 'acc = (fun _ _ -> self#zero)

    method private on_list : 'env 'a. ('env -> 'a -> 'acc) -> 'env -> 'a list -> 'acc =
      (fun f env xs -> self#list_fold_left f env self#zero xs)

    method private on_option : 'env 'a. ('env -> 'a -> 'acc) -> 'env -> 'a option -> 'acc =
      (fun f env -> Option.value_map ~default:self#zero ~f:(f env))

    method private list_fold_left
        : 'env 'a. ('env -> 'a -> 'acc) -> 'env -> 'acc -> 'a list -> 'acc =
      fun f env acc xs ->
        match xs with
        | [] -> acc
        | y :: ys ->
          let acc = self#plus acc (f env y) in
          self#list_fold_left f env acc ys
  end

class virtual ['self] mapreduce_ty_base =
  object (self : 'self)
    inherit ['acc] monoid

    method private on_string : 'env -> string -> string * 'acc = (fun _ x -> (x, self#zero))

    method private on_bool : 'env -> bool -> bool * 'acc = (fun _ x -> (x, self#zero))

    method private on_int : 'env -> int -> int * 'acc = (fun _ x -> (x, self#zero))

    method private on_symbol : 'env -> symbol -> symbol * 'acc = (fun _ x -> (x, self#zero))

    method private on_aloc : 'env -> ALoc.t -> ALoc.t * 'acc = (fun _ x -> (x, self#zero))

    method private on_list : 'a 'b. ('env -> 'a -> 'b * 'acc) -> 'env -> 'a list -> 'b list * 'acc
        =
      (fun f env -> self#list_fold_left f env ([], self#zero))

    method private on_option
        : 'a 'b. ('env -> 'a -> 'b * 'acc) -> 'env -> 'a option -> 'b option * 'acc =
      fun f env x ->
        match x with
        | None -> (None, self#zero)
        | Some x ->
          let (x', acc) = f env x in
          (Some x', acc)

    method private list_fold_left
        : 'a 'b. ('env -> 'a -> 'b * 'acc) -> 'env -> 'b list * 'acc -> 'a list -> 'b list * 'acc =
      fun f env acc xs ->
        match xs with
        | [] ->
          let (ys_rev, acc) = acc in
          (List.rev ys_rev, acc)
        | y :: ys ->
          let (acc_ys, acc) = acc in
          let (y', acc') = f env y in
          let acc'' = self#plus acc acc' in
          self#list_fold_left f env (y' :: acc_ys, acc'') ys
  end

File: rust_port/crates/flow_common_ty/src/ty_ancestors.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use flow_aloc::ALoc;
use flow_common::reason::Name;

use crate::ty_symbol::Symbol;

pub trait TyIterBase<Env, L> {
    fn on_string(&mut self, _env: &Env, _s: &str) {}

    fn on_name(&mut self, _env: &Env, _name: &Name) {}

    fn on_bool(&mut self, _env: &Env, _b: bool) {}

    fn on_int(&mut self, _env: &Env, _i: i64) {}

    fn on_symbol(&mut self, _env: &Env, _sym: &Symbol<L>) {}

    fn on_aloc(&mut self, _env: &Env, _loc: &ALoc) {}

    fn on_option<T, B, F>(&mut self, f: F, env: &Env, opt: &Option<T>) -> Option<B>
    where
        F: Fn(&mut Self, &Env, &T) -> B,
    {
        opt.as_ref().map(|x| f(self, env, x))
    }

    fn on_list<T, F>(&mut self, f: F, env: &Env, list: &[T])
    where
        F: Fn(&mut Self, &Env, &T),
    {
        for item in list {
            f(self, env, item);
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum StructuralMismatch {
    Mismatch,
    Difference(i32),
}

impl StructuralMismatch {
    pub fn assert0(n: i32) -> Result<(), StructuralMismatch> {
        if n == 0 {
            Ok(())
        } else {
            Err(StructuralMismatch::Difference(n))
        }
    }

    pub fn fail_gen<T>(tag_of: impl Fn(&T) -> i32, x: &T, y: &T) -> Result<(), StructuralMismatch> {
        Self::assert0(tag_of(x) - tag_of(y))
    }
}

pub trait TyIter2Base<Env, L> {
    fn on_string(&mut self, env: &Env, s1: &str, s2: &str) -> Result<(), StructuralMismatch> {
        /* In order to sort integer literals we try to parse all strings as integers */
        match s1.parse::<i64>() {
            Ok(xi) => match s2.parse::<i64>() {
                /* If both parse as integers then we compare them as integers */
                Ok(yi) => self.on_int(env, xi, yi),
                /* If xor parses as an integer then that one is "less than" the other */
                Err(_) => Err(StructuralMismatch::Difference(-1)),
            },
            Err(_) => match s2.parse::<i64>() {
                Ok(_) => Err(StructuralMismatch::Difference(1)),
                /* If neither parse as integers then we compare them as strings */
                Err(_) => StructuralMismatch::assert0(s1.cmp(s2) as i32),
            },
        }
    }

    fn on_name(&mut self, env: &Env, name1: &Name, name2: &Name) -> Result<(), StructuralMismatch> {
        self.on_string(env, name1.as_str(), name2.as_str())
    }

    fn on_bool(&mut self, _env: &Env, b1: bool, b2: bool) -> Result<(), StructuralMismatch> {
        StructuralMismatch::assert0(b1.cmp(&b2) as i32)
    }

    fn on_int(&mut self, _env: &Env, i1: i64, i2: i64) -> Result<(), StructuralMismatch> {
        StructuralMismatch::assert0((i1 - i2).signum() as i32)
    }

    fn on_symbol(
        &mut self,
        _env: &Env,
        sym1: &Symbol<L>,
        sym2: &Symbol<L>,
    ) -> Result<(), StructuralMismatch> {
        StructuralMismatch::assert0(sym1.sym_name.cmp(&sym2.sym_name) as i32)
    }

    fn on_aloc(&mut self, _env: &Env, loc1: &ALoc, loc2: &ALoc) -> Result<(), StructuralMismatch> {
        StructuralMismatch::assert0(loc1.cmp(loc2) as i32)
    }

    fn fail_option<T>(
        &mut self,
        _env: &Env,
        opt1: &Option<T>,
        _opt2: &Option<T>,
    ) -> Result<(), StructuralMismatch> {
        match opt1 {
            None => Err(StructuralMismatch::Difference(-1)),
            Some(_) => Err(StructuralMismatch::Difference(1)),
        }
    }

    fn on_option<T, F>(
        &mut self,
        f: F,
        env: &Env,
        opt1: &Option<T>,
        opt2: &Option<T>,
    ) -> Result<(), StructuralMismatch>
    where
        F: Fn(&mut Self, &Env, &T, &T) -> Result<(), StructuralMismatch>,
    {
        match (opt1, opt2) {
            (Some(x), Some(y)) => f(self, env, x, y),
            (None, None) => Ok(()),
            _ => self.fail_option(env, opt1, opt2),
        }
    }

    fn fail_list<T>(
        &mut self,
        _env: &Env,
        list1: &[T],
        _list2: &[T],
    ) -> Result<(), StructuralMismatch> {
        if list1.is_empty() {
            Err(StructuralMismatch::Difference(-1))
        } else {
            Err(StructuralMismatch::Difference(1))
        }
    }

    fn on_list<T, F>(
        &mut self,
        f: F,
        env: &Env,
        list1: &[T],
        list2: &[T],
    ) -> Result<(), StructuralMismatch>
    where
        F: Fn(&mut Self, &Env, &T, &T) -> Result<(), StructuralMismatch>,
    {
        if list1.len() != list2.len() {
            return self.fail_list(env, list1, list2);
        }
        for (item1, item2) in list1.iter().zip(list2.iter()) {
            f(self, env, item1, item2)?;
        }
        Ok(())
    }
}

pub trait TyMapBase<Env, L> {
    fn on_string(&mut self, _env: &Env, s: String) -> String {
        s
    }

    fn on_name(&mut self, _env: &Env, name: Name) -> Name {
        name
    }

    fn on_bool(&mut self, _env: &Env, b: bool) -> bool {
        b
    }

    fn on_int(&mut self, _env: &Env, i: i64) -> i64 {
        i
    }

    fn on_symbol(&mut self, _env: &Env, sym: Symbol<L>) -> Symbol<L> {
        sym
    }

    fn on_aloc(&mut self, _env: &Env, loc: ALoc) -> ALoc {
        loc
    }

    fn on_list<T: Clone, F>(&mut self, f: F, env: &Env, list: &[T]) -> Vec<T>
    where
        F: Fn(&mut Self, &Env, T) -> T,
    {
        list.iter().map(|x| f(self, env, x.clone())).collect()
    }

    fn on_option<T, F>(&mut self, f: F, env: &Env, opt: Option<T>) -> Option<T>
    where
        F: Fn(&mut Self, &Env, T) -> T,
    {
        opt.map(|x| f(self, env, x))
    }
}

pub trait TyEndoBase<Env, L> {
    fn on_string(&mut self, _env: &Env, s: String) -> String {
        s
    }

    fn on_name(&mut self, _env: &Env, name: Name) -> Name {
        name
    }

    fn on_bool(&mut self, _env: &Env, b: bool) -> bool {
        b
    }

    fn on_int(&mut self, _env: &Env, i: i64) -> i64 {
        i
    }

    fn on_symbol(&mut self, _env: &Env, sym: Symbol<L>) -> Symbol<L> {
        sym
    }

    fn on_aloc(&mut self, _env: &Env, loc: ALoc) -> ALoc {
        loc
    }

    /* Copied from
     * https://github.com/facebook/hhvm/blob/master/hphp/hack/src/ast/ast_defs_visitors_ancestors.ml
     */
    fn on_list<T: Clone, F>(&mut self, f: F, env: &Env, list: &[T]) -> Vec<T>
    where
        F: Fn(&mut Self, &Env, T) -> T,
    {
        list.iter().map(|v| f(self, env, v.clone())).collect()
    }

    fn on_option<T, F>(&mut self, f: F, env: &Env, opt: Option<T>) -> Option<T>
    where
        F: Fn(&mut Self, &Env, T) -> T,
    {
        opt.map(|x| f(self, env, x))
    }
}

pub trait Monoid {
    fn zero() -> Self;

    fn plus(a: Self, b: Self) -> Self;
}

pub trait TyReduceBase<Env, L> {
    type Acc: Monoid;

    fn on_string(&mut self, _env: &Env, _s: &str) -> Self::Acc {
        Self::Acc::zero()
    }

    fn on_name(&mut self, _env: &Env, _name: &Name) -> Self::Acc {
        Self::Acc::zero()
    }

    fn on_int(&mut self, _env: &Env, _i: i64) -> Self::Acc {
        Self::Acc::zero()
    }

    fn on_bool(&mut self, _env: &Env, _b: bool) -> Self::Acc {
        Self::Acc::zero()
    }

    fn on_symbol(&mut self, _env: &Env, _sym: &Symbol<L>) -> Self::Acc {
        Self::Acc::zero()
    }

    fn on_aloc(&mut self, _env: &Env, _loc: &ALoc) -> Self::Acc {
        Self::Acc::zero()
    }

    fn on_list<T, F>(&mut self, f: F, env: &Env, list: &[T]) -> Self::Acc
    where
        F: Fn(&mut Self, &Env, &T) -> Self::Acc,
    {
        self.list_fold_left(f, env, Self::Acc::zero(), list)
    }

    fn on_option<T, F>(&mut self, f: F, env: &Env, opt: &Option<T>) -> Self::Acc
    where
        F: Fn(&mut Self, &Env, &T) -> Self::Acc,
    {
        match opt {
            Some(x) => f(self, env, x),
            None => Self::Acc::zero(),
        }
    }

    fn list_fold_left<T, F>(&mut self, f: F, env: &Env, mut acc: Self::Acc, list: &[T]) -> Self::Acc
    where
        F: Fn(&mut Self, &Env, &T) -> Self::Acc,
    {
        for item in list {
            let item_acc = f(self, env, item);
            acc = Self::Acc::plus(acc, item_acc);
        }
        acc
    }
}

pub trait TyMapReduceBase<Env, L> {
    type Acc: Monoid;

    fn on_string(&mut self, _env: &Env, s: String) -> (String, Self::Acc) {
        (s, Self::Acc::zero())
    }

    fn on_name(&mut self, _env: &Env, name: Name) -> (Name, Self::Acc) {
        (name, Self::Acc::zero())
    }

    fn on_bool(&mut self, _env: &Env, b: bool) -> (bool, Self::Acc) {
        (b, Self::Acc::zero())
    }

    fn on_int(&mut self, _env: &Env, i: i64) -> (i64, Self::Acc) {
        (i, Self::Acc::zero())
    }

    fn on_symbol(&mut self, _env: &Env, sym: Symbol<L>) -> (Symbol<L>, Self::Acc) {
        (sym, Self::Acc::zero())
    }

    fn on_aloc(&mut self, _env: &Env, loc: ALoc) -> (ALoc, Self::Acc) {
        (loc, Self::Acc::zero())
    }

    fn on_list<T: Clone, U, F>(&mut self, f: F, env: &Env, list: &[T]) -> (Vec<U>, Self::Acc)
    where
        F: Fn(&mut Self, &Env, T) -> (U, Self::Acc),
    {
        self.list_fold_left(f, env, (Vec::new(), Self::Acc::zero()), list)
    }

    fn on_option<T, U, F>(&mut self, f: F, env: &Env, opt: Option<T>) -> (Option<U>, Self::Acc)
    where
        F: Fn(&mut Self, &Env, T) -> (U, Self::Acc),
    {
        match opt {
            Some(x) => {
                let (y, acc) = f(self, env, x);
                (Some(y), acc)
            }
            None => (None, Self::Acc::zero()),
        }
    }

    fn list_fold_left<T: Clone, U, F>(
        &mut self,
        f: F,
        env: &Env,
        acc: (Vec<U>, Self::Acc),
        list: &[T],
    ) -> (Vec<U>, Self::Acc)
    where
        F: Fn(&mut Self, &Env, T) -> (U, Self::Acc),
    {
        let (mut result, mut total_acc) = acc;
        for item in list {
            let (mapped, item_acc) = f(self, env, item.clone());
            result.push(mapped);
            total_acc = Self::Acc::plus(total_acc, item_acc);
        }
        (result, total_acc)
    }
}

impl Monoid for () {
    fn zero() -> Self {}

    fn plus(_a: Self, _b: Self) -> Self {}
}

impl Monoid for i64 {
    fn zero() -> Self {
        0
    }

    fn plus(a: Self, b: Self) -> Self {
        a + b
    }
}

impl Monoid for usize {
    fn zero() -> Self {
        0
    }

    fn plus(a: Self, b: Self) -> Self {
        a + b
    }
}

impl Monoid for bool {
    fn zero() -> Self {
        false
    }

    fn plus(a: Self, b: Self) -> Self {
        a || b
    }
}

impl<T> Monoid for Vec<T> {
    fn zero() -> Self {
        Vec::new()
    }

    fn plus(mut a: Self, b: Self) -> Self {
        a.extend(b);
        a
    }
}

impl<T> Monoid for Option<T> {
    fn zero() -> Self {
        None
    }

    fn plus(a: Self, b: Self) -> Self {
        a.or(b)
    }
}
